File: GP4patcher.py
import binascii
import hashlib


def sha1file():
    sha1h = hashlib.sha1()
    ba = bytearray(128*1024)
    mv = memoryview(ba)
    try:
        with open('GP4.exe', 'rb', buffering=0) as gp4:
            for n in iter(lambda: gp4.readinto(mv), 0):
                sha1h.update(mv[:n])
        return sha1h.hexdigest()
    except IOError:
        print(
            "Couldn't find GP4.exe!\nMake sure it's in the same folder as this tool.")


def main():
    print("This tool will patch Grand Prix 4 v1.02 cracked .exe to remove the CD Drive check\nthat requires you to have the ISO mounted despite it being cracked.")
    if (input("Proceed? Y/N\n").lower() == 'Y'.lower()):
        if (sha1file() == '592903176a7d52e0ad0deef2541a88554ed4181d'):
            try:
                print("Patching...")
                with open('GP4.exe', 'r+b') as gp4:
                    gp4.seek(0x0014bc0f)
                    gp4.write(binascii.unhexlify('9090'))
                    gp4.seek(0x0014bc18)
                    gp4.write(binascii.unhexlify('9090'))
                    print("Patch completed successfully.")
            except IOError:
                print(
                    "Couldn't find GP4.exe!\nMake sure it's in the same folder as this tool.")
        elif(sha1file() == '346e63c4ed47032e496a6ae90e6ff70109fb9529'):
            print("The file has already been patched.")
        else:
            print(
                f"Your GP4.exe is not the proper file. The supported file has SHA256 hash 592903176a7d52e0ad0deef2541a88554ed4181d\nYour file is {sha1file()}")
    else:
        print("Bye.")


if __name__ == '__main__':
    main()

File: GP4patcher.nim
import streams, strutils, std/sha1

proc calculateSHA1() : string =
    try:
        let hash = secureHashFile("GP4.exe")
        return $hash
    except IOError:
        echo "Couldn't open GP4.exe! Make sure it's in the same directory as this tool and that you have rights to that directory!"
        

proc main() =
    echo "This tool will patch the common Grand Prix 4 1.02 cracked exe to remove the requirement for a CD Drive."
    if calculateSHA1() == "592903176A7D52E0AD0DEEF2541A88554ED4181D":
        echo "Proceed? Y/N"
        let proceed = readLine(stdin)
        if proceed.toLowerAscii == "y":
            try:
                echo "Patching..."
                let gp4file = newFileStream("GP4.exe", fmReadWriteExisting)
                defer: gp4file.close
                gp4file.setPosition(0x0014bc0f)
                gp4file.write("\x90\x90")
                gp4file.setPosition(0x0014bc18)
                gp4file.write("\x90\x90")
                gp4file.close()
                echo "Patching complete. Press enter to quit."
                let quit = readLine(stdin)
            except IOError:
                echo "Couldn't open GP4.exe! Make sure it's in the same directory as this tool and that you have rights to that directory!"
        else:
            echo "Bye."
    elif calculateSHA1() == "346E63C4ED47032E496A6AE90E6FF70109FB9529":
        echo "The file has already been patched. Press enter to quit."
        let quit = readLine(stdin)
    else:
        echo "Your GP4.exe is not the one expected by this tool. Press enter to quit."
        let quit = readLine(stdin)

main()